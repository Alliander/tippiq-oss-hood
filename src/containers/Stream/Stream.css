@import "../../theme/variables/screen.css";

.title {
  margin: 6px 0 30px 0;
}

.warning {
  margin: 30px;
  text-align: center;
}

.navigationWrapper {
  max-width: 220px;
  width: 22.5%;
}

.wrapper {
  @media (--screen-sm-maxwidth) {
    background-color: #eeeeee;
    border-top: #dddddd;
    padding-top: 10px;
  }
  padding-bottom: 30px;
}

.gridHolder {
  position: inherit;
  width: 77.5%;
  padding-right: 0;

  @media (--screen-xs-maxwidth) {
    width: 100%;
    padding-right: 15px;
  }

  @media (--screen-sm-maxwidth) {
    width: 77.5%;
    padding-right: 15px;
  }
}

.alignWithBottom {
  position: relative;
  padding-top: 11px;
}
