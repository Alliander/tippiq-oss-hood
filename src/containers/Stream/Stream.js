/**
 * Stream container.
 * @module containers/Stream
 */

import React, { PropTypes, Component } from 'react';
import Helmet from 'react-helmet';
import { connect } from 'react-redux';
import {
  ButtonGroup,
  Col,
  Grid,
  Row,
} from 'react-bootstrap';
import { get } from 'lodash';

import { Footer, MasonryGrid, SharePopup } from '../../components';
import {
  DesktopSort,
  Header,
  MobileFilter,
  MobileSort,
  ServiceFilter,
  ThemeFilter,
} from '../../containers';

import {
  getCards,
  getPlaceLocation,
  sendFirstWeeklyNotification,
  initServiceFilter,
  toggleThemeFilter,
  toggleServiceFilter,
} from '../../actions';
import piwik from '../../piwik';

import styles from './Stream.css';

/**
 * Stream container.
 * @class Stream
 */
export class Stream extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    cards: PropTypes.object,
    placeLocation: PropTypes.object,
    userSession: PropTypes.object,
    frontendBaseUrl: PropTypes.string,
    location: PropTypes.object,
    themes: PropTypes.string,
    enabledThemes: PropTypes.array,
    services: PropTypes.string,
    enabledServices: PropTypes.array,
    allServices: PropTypes.array,
    filteredServices: PropTypes.array,
    getPlaceLocation: PropTypes.func.isRequired,
    getCards: PropTypes.func.isRequired,
    sendFirstWeeklyNotification: PropTypes.func.isRequired,
    initServiceFilter: PropTypes.func.isRequired,
    toggleThemeFilter: PropTypes.func.isRequired,
    toggleServiceFilter: PropTypes.func.isRequired,
    router: PropTypes.object,
  }

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Properties
   * @returns {undefined}
   */
  constructor(props) {
    super(props);
    this.state = {
      shareButtonEl: null,
      sharePopupVisible: false,
    };
    this.handleOnShowSharePopup = this.handleOnShowSharePopup.bind(this);
    this.renderSharePopup = this.renderSharePopup.bind(this);
    this.hidePopup = this.hidePopup.bind(this);
  }

  /**
   * On component did mount
   * @function componentDidMount
   * @returns {undefined}
   */
  componentDidMount() {
    this.initLocation();
    const emailIsVerified = get(this.props, 'location.query.emailIsVerified');
    if (emailIsVerified) {
      this.props.sendFirstWeeklyNotification(this.props.userSession.placeId);
    }
    const track = get(this.props, 'location.query.track');
    if (track) {
      piwik.push(['trackEvent', 'Register flow', `Land on stream (step 3) - track: ${track}`]);
    }
  }

  /**
   * Component will receive props
   * @method componentWillReceiveProps
   * @param {Object} nextProps Next properties
   * @returns {undefined}
   */
  componentWillReceiveProps(nextProps) {
    const { userSession, placeLocation, allServices } = nextProps;
    if (userSession.loggedIn && placeLocation.idle) {
      this.props.getPlaceLocation(userSession.placeId);
    } else if (!this.filtersInitialized && placeLocation.location && allServices.length > 0) {
      // init active filters from route after location & all services are loaded
      this.initFilters(nextProps);
    }
    if (this.filtersInitialized) {
      this.updateRoute(nextProps);
    }
  }

  /**
   * Init filters
   * @method initFilters
   * @param {Object} props properties
   * @returns {undefined}
   */
  initFilters(props) {
    const { themes, services, cards, placeLocation } = props;
    const startThemes = themes ? themes.split(',') : [];
    const startServices = services ? services.split(',') : [];
    if (startThemes.length > 0 || startServices.length > 0) {
      props.initServiceFilter(startThemes, startServices);
    } else {
      props.getCards(cards.sort, placeLocation.location.geometry);
    }
    this.filtersInitialized = true;
  }

  /**
   * Update route with active filters
   * @param {Object} props properties
   * @method updateRoute
   * @returns {undefined}
   */
  updateRoute(props) {
    const { enabledThemes, enabledServices, filteredServices, router } = props;
    const myServices = filteredServices.length === enabledServices.length ? [] : enabledServices;

    const pathname = `/mijn-buurt${enabledThemes.length > 0 ? `/themas/${enabledThemes.join(',')}`
      : ''}${myServices.length > 0 ? `/diensten/${myServices.join(',')}` : ''}`;
    if (pathname !== decodeURI(router.getCurrentLocation().pathname)) {
      router.replace({ pathname });
    }
  }

  /**
   * Init location
   * @method initLocation
   * @returns {undefined}
   */
  initLocation() {
    const { userSession, cards, placeLocation } = this.props;
    if (placeLocation.location && placeLocation.location.geometry) {
      this.props.getCards(cards.sort, placeLocation.location.geometry);
    } else if (userSession.placePolicyUrl) {
      window.location.href = userSession.placePolicyUrl;
    } else if (userSession.placeId) {
      this.props.getPlaceLocation(userSession.placeId);
    }
  }
  /**
   * Handle onFlip method.
   * @method handleOnShowSharePopup
   * @param {Object} shareButtonEl element
   * @returns {undefined}
   */
  handleOnShowSharePopup(shareButtonEl, { id, ctaLabel }) {
    if (shareButtonEl !== this.state.shareButtonEl) {
      this.setState({ sharePopupVisible: false }, () => {
        this.setState({
          sharePopupVisible: true,
          shareButtonEl,
          shareData: {
            id,
            ctaLabel,
          },
        });
      });
    } else {
      this.hidePopup();
    }
  }

  /**
   * Hide Popup.
   * @method hidePopup
   * @returns {undefined}
   */
  hidePopup() {
    this.setState({ sharePopupVisible: false, shareButtonEl: null, shareData: {} });
  }

  /**
   * SharePopup.
   * @method renderSharePopup
   * @returns {string} Markup for the component.
   */
  renderSharePopup() {
    const { frontendBaseUrl } = this.props;

    return (
      <SharePopup
        frontendBaseUrl={frontendBaseUrl}
        buttons={[
          {
            icon: 'facebook',
          },
          {
            icon: 'twitter',
          },
          {
            icon: 'whatsapp',
          },
          {
            icon: 'mail',
          },
        ]}
        shareData={this.state.shareData}
        linkElement={this.state.shareButtonEl}
      />
    );
  }

  /**
   * Render cards view.
   * @function renderCardsView
   * @returns {string} Markup of the container.
   */
  renderCardsView() {
    const { cards, placeLocation } = this.props;
    let cardsView;

    if (placeLocation.placePolicyUrl) {
      cardsView = (
        <div className={styles.warning}>
          We hebben geen toegang tot je locatie en kunnen daardoor geen kaartjes tonen.<br />
          <a href={placeLocation.placePolicyUrl}>Klik hier om je Huisregels te beheren</a>
        </div>
      );
    } else {
      cardsView = !cards.pending && !cards.idle && cards.items.length === 0 ?
        <div className={styles.warning}>
          Er zijn geen kaartjes gevonden in jouw buurt met de geselecteerde zoekcritera.
        </div>
        :
        <MasonryGrid
          handleOnShowSharePopup={this.handleOnShowSharePopup}
          items={cards.items}
        />;
    }

    return (
      <Col sm={9} className={styles.gridHolder}>
        {this.state.sharePopupVisible && this.renderSharePopup()}
        {!placeLocation.error ? cardsView :
          <div className={styles.warning}>
            We hebben geen toegang tot je locatie en kunnen daardoor geen kaartjes tonen.
          </div>
        }
      </Col>
    );
  }

  /**
   * renders.
   * @function renders
   * @returns {string} Markup of the container.
   */
  render() {
    return (
      <div id="page-stream">
        <Helmet title="Dit gebeurt er in je buurt" />
        <Header />
        <Grid className={styles.wrapper}>
          <Row>
            <Col sm={12}>
              <div className="visible-xs">
                <ButtonGroup>
                  <MobileFilter />
                  <MobileSort />
                </ButtonGroup>
              </div>
              <div className={`pull-right hidden-xs ${styles.alignWithBottom}`}>
                <DesktopSort />
              </div>
              <h1 className={styles.title}>Dit gebeurt er in je buurt</h1>
            </Col>
          </Row>
          <Row>
            <Col sm={3} className={`hidden-xs ${styles.navigationWrapper}`}>
              <ThemeFilter />
              <ServiceFilter />
            </Col>
            {this.renderCardsView()}
          </Row>
        </Grid>
        <Footer />
      </div>
    );
  }
}

export default connect(
  (state, props) => ({
    placeLocation: state.placeLocation,
    cards: state.cards,
    frontendBaseUrl: state.appConfig.frontendBaseUrl,
    userSession: state.userSession,
    themes: props.params.themes,
    services: props.params.services,
    enabledThemes: state.services.enabledThemes,
    enabledServices: state.services.enabledServices,
    allServices: state.services.services,
    filteredServices: state.services.filteredServices,
  }),
  ({
    getCards,
    getPlaceLocation,
    sendFirstWeeklyNotification,
    initServiceFilter,
    toggleThemeFilter,
    toggleServiceFilter,
  })
)(Stream);
