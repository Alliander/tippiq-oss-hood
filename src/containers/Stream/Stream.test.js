import React from 'react';
import renderer from 'react-test-renderer';
import { Stream } from './Stream';

jest.mock('../../piwik', () => ({
  push: jest.fn(),
}));

jest.mock('react-masonry-component');
jest.mock('../Header/Header.js',
  () => jest.fn(() => <div />));
jest.mock('../ThemeFilter/ThemeFilter.js',
  () => jest.fn(() => <div />));
jest.mock('../ServiceFilter/ServiceFilter.js',
  () => jest.fn(() => <div />));
jest.mock('../DesktopSort/DesktopSort.js',
  () => jest.fn(() => <div />));
jest.mock('../MobileSort/MobileSort.js',
  () => jest.fn(() => <div />));

// https://github.com/reactjs/redux/issues/588 (recommended method below)
// http://ericnish.io/blog/how-to-unit-test-react-redux-components/

test('Renders', () => {
  // This is where we create a mock state/props
  const props = {
    cards: {
      items: [],
    },
    placeLocation: {
      location: {},
    },
    userSession: {},
    getCards: () => {},
    getPlaceLocation: () => {},
  };

  const component = renderer.create(
    <Stream {...props} />
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
