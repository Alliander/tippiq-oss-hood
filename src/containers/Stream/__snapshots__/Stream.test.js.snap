exports[`test Renders 1`] = `
<div
  id="page-stream">
  <div />
  <div
    className="container">
    <div
      className="row">
      <div
        className="col-sm-12">
        <div
          className="visible-xs">
          <div
            className="btn-group">
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button">
              Filteren
            </button>
            <div />
          </div>
        </div>
        <div
          className="pull-right hidden-xs undefined">
          <div />
        </div>
        <h1
          className={undefined}>
          Dit gebeurt er in je buurt
        </h1>
      </div>
    </div>
    <div
      className="row">
      <div
        className="hidden-xs undefined col-sm-3">
        <div />
        <div />
      </div>
      <div
        className="col-sm-9">
        <div
          className={undefined}>
          Er zijn geen kaartjes gevonden in jouw buurt met de geselecteerde zoekcritera.
        </div>
      </div>
    </div>
  </div>
  <div
    className="container-fluid footer undefined">
    <div
      className="hidden-xs">
      <a
        className={undefined}
        onClick={[Function]}
        style={Object {}}>
        Over Tippiq Buurt
      </a>
      <a
        className={undefined}
        onClick={[Function]}
        style={Object {}}>
        Partners
      </a>
      <a
        className={undefined}
        onClick={[Function]}
        style={Object {}}>
        Privacy & Disclaimer
      </a>
      <a
        className={undefined}
        onClick={[Function]}
        style={Object {}}>
        Contact
      </a>
    </div>
    <div
      className={undefined}>
      <a
        className={undefined}
        href="https://www.facebook.com/Tippiq-1430219107214646">
        <i
          className="fa fa-facebook" />
      </a>
      <a
        className={undefined}
        href="https://twitter.com/tippiq">
        <i
          className="fa fa-twitter" />
      </a>
    </div>
    <div
      className={undefined}>
      Copyright Â© 2017 Alliander N.V.
    </div>
    <nav
      className="visible-xs undefined navbar navbar-default">
      <div
        className="container">
        <div
          className="nav navbar-nav">
          <li
            className="visible-xs dropdown">
            <a
              aria-expanded={false}
              aria-haspopup={true}
              className="dropdown-toggle"
              href="#"
              id="footer"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button">
              Meer over Tippiq
               
              <span
                className="caret" />
            </a>
            <ul
              aria-labelledby="footer"
              className="dropdown-menu"
              role="menu">
              <li
                className=""
                role="presentation"
                style={undefined}>
                <a
                  action="push"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1">
                  Over Tippiq Buurt
                </a>
              </li>
              <li
                className=""
                role="presentation"
                style={undefined}>
                <a
                  action="push"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1">
                  Partners
                </a>
              </li>
              <li
                className=""
                role="presentation"
                style={undefined}>
                <a
                  action="push"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1">
                  Privacy & Disclaimer
                </a>
              </li>
              <li
                className=""
                role="presentation"
                style={undefined}>
                <a
                  action="push"
                  href="#"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1">
                  Contact
                </a>
              </li>
            </ul>
          </li>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
